File: 3264-final-array-state-after-k-multiplication-operations-i/3264-final-array-state-after-k-multiplication-operations-i.cpp
class Solution {
public:
    vector<int> getFinalState(vector<int>& nums, int k, int multiplier) {
        vector<pair<int, int>> heap(nums.size());
        for (int i = 0; i < nums.size(); i++) 
            heap[i] = {nums[i], i};
        
        make_heap(heap.begin(), heap.end(), greater<>());
        
        while (k--) {
            pop_heap(heap.begin(), heap.end(), greater<>());
            auto [num, idx] = heap.back();
            heap.pop_back();
            nums[idx] = num * multiplier;
            heap.push_back({nums[idx], idx});
            push_heap(heap.begin(), heap.end(), greater<>());
        }
        return nums;
    }
};

File: 3264-final-array-state-after-k-multiplication-operations-i/3264-final-array-state-after-k-multiplication-operations-i.py
import heapq

class Solution:
    def getFinalState(self, nums: list[int], k: int, multiplier: int) -> list[int]:
        heap = [(num, i) for i, num in enumerate(nums)]
        heapq.heapify(heap)  # Convert to min-heap

        for _ in range(k):
            num, idx = heapq.heappop(heap)  # Pop the smallest element
            nums[idx] = num * multiplier   # Update the value in nums
            heapq.heappush(heap, (nums[idx], idx))  # Push the updated pair back into the heap

        return nums
